packages/block-library/src/post-author/edit.js

```diff
@@ -1,31 +1,255 @@
+
+	const { author, authors } = useSelect(
+		( select ) => {
+			const { getEntityRecord, getAuthors } = select( 'core' );
+			return {
+				author: getEntityRecord( 'root', 'user', postAuthorId ),
+				authors: getAuthors(),
+			};
+		},
+		[ postAuthorId ]
+	);
+
+	const { isSelected, fontSize, setFontSize } = props;
+	const {
+		TextColor,
+		BackgroundColor,
+		InspectorControlsColorPanel,
+		ColorPanel,
+	} = __experimentalUseColors(
+		[
+			{ name: 'textColor', property: 'color' },
+			{ name: 'backgroundColor', className: 'background-color' },
+		],
+		{
+			contrastCheckers: [
+				{
+					backgroundColor: true,
+					textColor: true,
+					fontSize: fontSize.size,
+				},
+			],
+			colorDetector: { targetRef: ref },
+			colorPanelProps: {
+				initialOpen: true,
+			},
+		},
+		[ fontSize.size ]
+	);
+
+	const { align, id, showAvatar, showBio, byline } = props.attributes;
+
+	const avatarSizes = [
+		{ value: 24, label: __( 'Small' ) },
+		{ value: 48, label: __( 'Medium' ) },
+		{ value: 96, label: __( 'Large' ) },
+	];
+
+	let avatarSize = DEFAULT_AVATAR_SIZE;
+	if ( !! props.attributes.avatarSize ) {
+		avatarSize = props.attributes.avatarSize;
+	}
+	useEffect( () => {
+		if ( author && ! props.attributes.id ) {
+			props.setAttributes( {
+				id: Number( author.id ),
+				name: author.name,
+				description: author.description,
+				avatarSize,
+				avatarUrl: author.avatar_urls[ avatarSize ],
+			} );
+		} else if ( authorId ) {
+			const authorFetch = apiFetch( {
+				path: '/wp/v2/users/' + authorId + '?context=edit',
+			} );
+			authorFetch.then( ( newAuthor ) => {
+				props.setAttributes( {
+					id: newAuthor.id,
+					name: newAuthor.name,
+					avatarUrl:
+						newAuthor.avatar_urls[ props.attributes.avatarSize ],
+					description: newAuthor.description,
+				} );
+			} );
+			return () => {
+				if ( authorFetch ) {
+					invoke( authorFetch, [ 'abort' ] );
+				}
+			};
+		}
+	}, [ authorId, author ] );
+
+	const blockClassNames = classnames( 'wp-block-post-author', {
+		[ fontSize.class ]: fontSize.class,
+	} );
+	const blockInlineStyles = {
+		fontSize: fontSize.size ? fontSize.size + 'px' : undefined,
+	};
+
+	return (
+		<>
+			<InspectorControls>
+				<PanelBody title={ __( 'Author Settings' ) }>
+					<SelectControl
+						label={ __( 'Author' ) }
+						value={ id }
+						options={ authors.map( ( theAuthor ) => {
+							return {
+								value: theAuthor.id,
+								label: theAuthor.name,
+							};
+						} ) }
+						onChange={ ( newAuthorId ) => {
+							setAuthorId( newAuthorId );
+						} }
+					/>
+					<ToggleControl
+						label={ __( 'Show avatar' ) }
+						checked={ showAvatar }
+						onChange={ () =>
+							props.setAttributes( { showAvatar: ! showAvatar } )
+						}
+					/>
+					{ showAvatar && (
+						<SelectControl
+							label={ __( 'Avatar size' ) }
+							value={ props.attributes.avatarSize }
+							options={ avatarSizes }
```

Comment:
```suggestion
							options={ Object.keys( author.avatur_urls ) }
```

(see https://github.com/WordPress/gutenberg/pull/19894/files#r413230270)

Reply:
> @ockham 1st thanks for the review ;) I went through most of the things you found.
> Regarding the refactoring of the author id, I do wonder how to proceed.
> 
> * there are an edit and a display component to avoid some `if`s because rules of hooks will not allow us to bail out early enough. E.g. we'd need to treat empty `postAuthorId` [here](https://github.com/WordPress/gutenberg/pull/19894/files#diff-7e82f718ff736f624069097db0995cb1R33)

Right, that makes sense. I think we have a few different options here:

1. Keep the `useEntityProp` call in `PostAuthorEdit`, but also get a setter and pass both the post author ID and the setter to the wrapped `PostAuthorDisplay`.
1. Depending on the next point below, we might end up needing fewer hooks (e.g. the `useEffect` might become obsolete, as might some component state), in which case we might be able to merge both components more easily.

> * we need the block to have an attribute for storing the author id because the block is not alywas synced to the post's id and we do not want to set the posts author from this block's settings. At least not with this UI.

Maybe we can reconsider that though? There seems to be a pattern with other `Post...` components that they actually allow changing the corresponding post values -- e.g. `PostDate` [actually sets the post's date](https://github.com/WordPress/gutenberg/blob/71d03197cf257dd4ddc4706e1efa788d884d1183/packages/block-library/src/post-date/edit.js#L60) (rather than an intermediate block attribute). For consistency's sake, I think that would also make sense for the `PostAuthor` block, and from a quick glance, I didn't actually see any big conceptual hurdles. But maybe I'm missing something? Curious to hear your thoughts @draganescu :slightly_smiling_face: 

> So, I'll see how to at least not need internal component state, if that's possible.

Sounds good for a start!